packages/engine/Source/Scene/Model/PntsLoader.js

```diff
@@ -572,6 +582,383 @@ function makeStructuralMetadata(parsedContent, customAttributeOutput) {
+    pivot_index = Math.floor((lo + hi) / 2);
+  }
+  let temp = A[hi];
+  A[hi] = pivot;
+  A[pivot_index] = temp;
+  let i = lo;
+  for (let j = lo; j < hi; j++) {
+    if (positions[3 * A[j] + axis] <= positions[3 * pivot + axis]) {
+      const tmp = A[i];
+      A[i] = A[j];
+      A[j] = tmp;
+      i += 1;
+    }
+  }
+  temp = A[hi];
+  A[hi] = A[i];
+  A[i] = temp;
+  return i;
+}
+
+function distance(v1, v2) {
+  return Math.sqrt(
+    (v1[0] - v2[0]) ** 2 + (v1[1] - v2[1]) ** 2 + (v1[2] - v2[2]) ** 2
+  );
+}
+
+function normalize(v) {
+  const newVector = new Float64Array(3);
+  const denom = Math.sqrt(v[0] ** 2 + v[1] ** 2 + v[2] ** 2);
+  newVector[0] = v[0] / denom;
+  newVector[1] = v[1] / denom;
+  newVector[2] = v[2] / denom;
+  return newVector;
+}
+
+/**
+ * Returns the t-value of the closest point on a line to a point.
+ * @param point
+ * @param a + t*b
+ * @returns t minimizing |point - (a + t*b)|
+ */
+function closestPointOnLineToPoint(point, a, b) {
+  const dir = normalize(b);
+  return (
+    dir[0] * (point[0] - a[0]) +
+    dir[1] * (point[1] - a[1]) +
+    dir[2] * (point[2] - a[2])
+  );
+}
+
+function closestPointOnRayToPoint(point, a, b) {
+  const t = closestPointOnLineToPoint(point, a, b);
+  if (t < 0.0) {
+    return a;
+  }
+  const intersection = new Float64Array(3);
+  intersection[0] = a[0] + t * b[0];
+  intersection[1] = a[1] + t * b[1];
+  intersection[2] = a[2] + t * b[2];
+  return intersection;
+}
+
+function distanceFromRayToPoint(ray, point) {
+  const origin = new Float64Array(3);
+  const dir = new Float64Array(3);
+  origin[0] = ray.origin.x;
+  origin[1] = ray.origin.y;
+  origin[2] = ray.origin.z;
+  dir[0] = ray.direction.x;
+  dir[1] = ray.direction.y;
+  dir[2] = ray.direction.z;
+  return distance(point, closestPointOnRayToPoint(point, origin, dir));
+}
+
+function boundingSphereOfBox(minx, maxx, miny, maxy, minz, maxz) {
+  const center = new Float64Array(3);
+  center[0] = (minx + maxx) / 2;
+  center[1] = (miny + maxy) / 2;
+  center[2] = (minz + maxz) / 2;
+  const vertex = new Float64Array(3);
+  vertex[0] = minx;
+  vertex[1] = miny;
+  vertex[2] = minz;
+  const radius = distance(center, vertex);
+  return [center, radius];
+}
+
+PntsLoader.prototype.findPointsWithinRadiusOfRay = function (ray, radius) {
+  const positions = this._enuCoords;
+  const indexedTree = this._3DTree;
+  const box = new Float64Array(6);
+  box[0] = this._boxENU[0];
+  box[1] = this._boxENU[1];
+  box[2] = this._boxENU[2];
+  box[3] = this._boxENU[3];
+  box[4] = this._boxENU[4];
+  box[5] = this._boxENU[5];
+  // bfsPrint(indexedTree, positions, box);
+  //console.log(box);
+  const curNode = this._beginNode;
+  const depth = 0;
+  const result = rayIterate(
+    indexedTree,
+    positions,
+    ray,
+    box,
+    curNode,
+    depth,
+    radius
+  );
+  if (!result) {
+    return null;
+  }
+  const ecefTransformationMatrix = Transforms.eastNorthUpToFixedFrame(
+    this._ecefRefPoint
```

Comment:
oh another comment: it's not ideal that this function takes in ENU and outputs ECEF; it's easier to reason about if the input and output are in the same frames, so I think you should return the output in tile-frame ENU and then transform back into ECEF in mach9-cloud

Reply:
The input for this function is actually in Scene.js which takes in a standard ECEF ray which is used for most cases; this is just the internal conversion back to ECEF since we lose the reference frame when we take out scope to the scene (since the scene contains multiple tiles and each tile has a different reference point), If we don't do the conversion here we have to create a whole new function in mach9-cloud that exposes the scene and tiles within it, loops through all the tiles and do the conversion for each tile, which is slower and not consistent with the picking code that it already written in mach9-cloud